File: core/src/main/java/com/geometryduel/game/state/PlayerState.java
package com.geometryduel.game.state;

import com.geometryduel.game.actor.PlayerActor;
import com.geometryduel.render.Shapes;

/** 还原 PlayerActorState：玩家动作状态机基类。 */
public abstract class PlayerState {
    public MoveState moveState;

    public abstract void act(PlayerActor p);

    public abstract void update(PlayerActor p);

    public abstract PlayerState entryState(PlayerActor p);

    public abstract void displayEffect(Shapes s, PlayerActor p);

    public boolean isDamaged() {
        return false;
    }

    public boolean isDrawingLongBow() {
        return false;
    }

    public boolean hasCompletedLongBowCharge(PlayerActor p) {
        return false;
    }

    /** 敌方玩家相对本玩家的角度（getEnemyPlayerActorAngle）。 */
    protected float enemyAngle(PlayerActor p) {
        PlayerActor enemy = p.group.enemyGroup.firstPlayer();
        if (enemy == null) return p.aimAngle;
        return (float) Math.atan2(enemy.pos.y - p.pos.y, enemy.pos.x - p.pos.x);
    }
}

File: app/src/main/java/com/geometryduel/game/state/PlayerState.kt
package com.geometryduel.game.state

import com.geometryduel.game.actor.PlayerActor
import com.geometryduel.render.GameRenderer

/** 还原 PlayerActorState：玩家动作状态机基类。 */
abstract class PlayerState {
    lateinit var moveState: MoveState

    abstract fun act(p: PlayerActor)

    abstract fun update(p: PlayerActor)

    abstract fun entryState(p: PlayerActor): PlayerState

    abstract fun displayEffect(s: GameRenderer, p: PlayerActor)

    open fun isDamaged() = false

    open fun isDrawingLongBow() = false

    open fun hasCompletedLongBowCharge(p: PlayerActor) = false

    /** 敌方玩家相对本玩家的角度（getEnemyPlayerActorAngle）。 */
    protected fun enemyAngle(p: PlayerActor): Float {
        val enemy = p.group?.enemyGroup?.firstPlayer() ?: return p.aimAngle
        return kotlin.math.atan2(enemy.pos.y - p.pos.y, enemy.pos.x - p.pos.x)
    }
}
